### PR TITLE
feat: add audio message

### PR DESCRIPTION
**Pull request by MihailMatveichuk**
- add audio message template
- coonect audio with audio player
Screenshot_1
Screenshot_2
Screenshot_3

**Send voice message beforehead**

### CODE REVIEW

File: src/components/Messanger/Footer/Footer.tsx [hunk 1]
```diff
@@ -14,12 +14,14 @@ interface FooterProps {
   selectedMessageText: string;
   showReplyMessage: boolean;
   onShowReplyMessage: () => void;
+  setBlobSrc: React.Dispatch<React.SetStateAction<string>>;
```
Comment (akrutsko) [inline src/components/Messanger/Footer/Footer.tsx:17]:
```suggestion
  setBlobSrc: Dispatch<SetStateAction<string>>;
```

File: src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx [hunk 2]
```diff
@@ -13,11 +13,13 @@ import { ModalCommon } from 'components/common';
 interface RecordingAudioProps {
   setIsRecording: React.Dispatch<React.SetStateAction<boolean>>;
   setIsRecordedBlob: React.Dispatch<React.SetStateAction<boolean>>;
+  setBlobSrc: React.Dispatch<React.SetStateAction<string>>;
```
Comment (akrutsko) [inline src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx:16]:
```suggestion
  setBlobSrc: Dispatch<SetStateAction<string>>;
```

File: src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import React, { useCallback, useRef, useState } from 'react';
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx:1]: get rid of React

File: src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import styles from './audioMessageContent.module.css';
+import css from '../message.module.css';
+import { ReactComponent as PlayIcon } from '../../../../../assets/Messanger/RecordingAudio/PlayIcon.svg';
+import { ReactComponent as PauseIcon } from '../../../../../assets/Messanger/RecordingAudio/PauseIcon.svg';
+import { ReactComponent as PinnedIcon } from '../../../../../assets/Messanger/iconForPinnedMessage.svg';
+import { formatTime } from './lib';
+
+type Props = {
+  audioRef: React.RefObject<HTMLAudioElement>;
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx:12]:
```suggestion
  audioRef: RefObject<HTMLAudioElement>;
```
I never used RefObject before :)

File: src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    cursorWidth: 0,
+    barGap: 1,
+    barWidth: 2,
+    normalize: true,
+  });
+
+  const onPlayPause = useCallback(() => {
+    wavesurfer && wavesurfer.playPause();
+  }, [wavesurfer]);
+  const handleAudioPlay = () => {
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx:41]: I would add an empty line...

File: src/components/Messanger/Workspace/Message/AudioMessageContent/audioMessageContent.module.css [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  gap: 2px;
+}
+
+.lowPart {
+  display: flex;
+  justify-content: space-between;
+}
+
+.audioDuration {
+  font-family: Montserrat;
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/AudioMessageContent/audioMessageContent.module.css:32]:
```suggestion
  font-family: var(--ff);
```

File: src/components/Messanger/Workspace/Message/AudioMessageContent/lib.ts
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/AudioMessageContent/lib.ts:1]: may be need to have a separate folder/file with all project utils? It seems this lib doesn't directly relate to audio message.

File: src/components/Messanger/Workspace/Message/Message.tsx [hunk 7]
```diff
@@ -60,11 +66,23 @@ export const Message: React.FC<MessageProps> = ({
...
+            <AudioMessageContent
+              audioRef={{ current: null }}
+              timestamp={timestamp}
+              isPinned={isPinned}
+              blob={blob}
+            />
+          ) : (
+            <>
+              <p className={css.messageText}>{text}</p>
+
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Message/Message.tsx:79]: empty line?

File: src/components/Messanger/Workspace/Workspace.tsx [hunk 8]
```diff
@@ -213,20 +233,21 @@ export const Workspace: FC<WorkspaceProps> = ({
...
+                type={message.type}
                 text={message.text}
                 isMe={message.isMe}
                 timestamp={message.timestamp}
                 isSelected={showContextMenu && selectedMessageId === message.id}
                 isPinned={pinnedMessages.some(
                   (pinnedMessage) => pinnedMessage.id === message.id,
                 )}
-                onContextMenu={(event) => {
+                onContextMenu={(event: React.MouseEvent<HTMLDivElement>) => {
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/Workspace.tsx:250]:
```suggestion
                onContextMenu={(event: MouseEvent<HTMLDivElement>) => {
```

File: src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx [hunk 9]
```diff
@@ -10,24 +10,30 @@ import { ReactComponent as PauseIcon } from '../../../../assets/AudioPlayer/Paus
...
 
+  const { speed, volume, duration, isPlaying } = useAppSelector(getAudioInfo);
+  const dispatch = useAppDispatch();
+
   useEffect(() => {
     const rangeInput = volumeRef.current as HTMLInputElement;
 
     if (rangeInput) {
-      rangeInput.style.setProperty('--thumb-percentage', `${volume * 100}%`);
+      rangeInput.style.setProperty('--thumb-percentage', `${volume! * 100}%`);
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx:36]: why do we need an exclamation mark here?

File: src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx [hunk 10]
```diff
@@ -63,11 +69,19 @@ export const AudioPlayer = () => {
     }
   };
 
+  const handlePlayPause = () => {
+    dispatch(audioInfoUpdate({ isPlaying: isPlaying === true ? false : true }));
```
Comment (akrutsko) [inline src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx:73]:
```suggestion
    dispatch(audioInfoUpdate({ isPlaying: !isPlaying }));
```

File: src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx (hunk 5 above)
Comment (MihailMatveichuk) [reply src/components/Messanger/Workspace/Message/AudioMessageContent/AudioMessageContent.tsx:53]: prettier doesnt allow formatting code

File: src/components/Messanger/Workspace/Message/AudioMessageContent/audioMessageContent.module.css (hunk 6 above)
Comment (MihailMatveichuk) [reply src/components/Messanger/Workspace/Message/AudioMessageContent/audioMessageContent.module.css:32]: done

File: src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx (hunk 10 above)
Comment (MihailMatveichuk) [reply src/components/Messanger/Workspace/AudioPlayer/AudioPlayer.tsx]: done

File: src/components/Messanger/Footer/Footer.tsx [hunk 11]
```diff
@@ -1,46 +1,32 @@
...
-import React, { useState, FC } from 'react';
+import { useState, ChangeEvent, Dispatch, SetStateAction } from 'react';
 import { ReactComponent as EmojiIcon } from '../../../assets/Messanger/emojiIcon.svg';
 import { ReactComponent as Clip } from '../../../assets/Messanger/Clip.svg';
-import EmojiPicker, { EmojiClickData } from 'emoji-picker-react';
 import { ReplyMessage } from './ReplyMessage';
 import { UploadButton } from 'components/pages/Questionnaire/UploadButton';
 import { ModalClipElements } from './ModalClipElements';
 import { RecordingAudio } from './RecordingAudio/RecordingAudio';
-import { useVoiceVisualizer, VoiceVisualizer } from 'react-voice-visualizer';
```
Comment (Sergei181294) [inline src/components/Messanger/Footer/Footer.tsx:11]: удали пакет из проекта чтоб не засорять)

File: src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx [hunk 12]
```diff
@@ -74,6 +75,14 @@ export const RecordingAudio: React.FC<RecordingAudioProps> = ({
     console.log('onResumeAudioPlayback');
   };
 
+  const sendRecordBlob = () => {
+    console.log('recordedBlob', audioSrc);
```
Comment (Sergei181294) [inline src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx:79]: delete)

File: src/components/Messanger/Footer/Footer.tsx (hunk 11 above)
Comment (MihailMatveichuk) [reply src/components/Messanger/Footer/Footer.tsx:11]: Согласен

File: src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx (hunk 12 above)
Comment (akrutsko) [inline src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx:79]: do we still need this console.log?

Comment (MihailMatveichuk) [reply src/components/Messanger/Footer/RecordingAudio/RecordingAudio.tsx:79]: Yes

**Pull request merged**